### PR TITLE
EndDate Expiring Schedule Introduction

### PR DESCRIPTION
**Pull request by kunkun-tang**
Today, when we specify a flow schedule in Azkaban, the flow will run infinitely without termination as per the schedule. As #721 proposes, should schedule have an end date, the flow will have more flexibilities to be managed. In this PR, I implement the end date feature to allow users to specify the expiration date given a schedule. UI has not been changed yet in this patch (will not be implemented in near future), but people will be able to call API to leverage the new feature.

The API command sample:
`curl -k -d ajax=scheduleCronFlow -d projectName=wtwt -d flow=azkaban -d endSchedTime=14XXXXXXXXXX --data-urlencode cronExpression="0 * * ? * 6" -b "azkaban.browser.session.id=XXXXXXXXXXXXXX" http://localhost:8081/schedule`

Similar to what @inoviavenkat implemented, I add an EndTimeChecker, as part of Expire Condition checker. Also, Added one parameter `endSchedTime` to schedule API. If the users don't specify this parameter, Azkaban would use the default value `-1`, and the schedule will be still running infinitely. Besides, I add Unit tests as much as possible to guarantee that this change will not break out. 

Furthermore, I made comprehensive manual tests and am glad to share:
* We have flow1, flow2, flow3, flow4, corresponding schedule1, schedule2, schedule3, schedule4. Every schedule runs flow every minute. The 4 flows keep undispatched at this stage.
* Ignoring this PR, I started a web server running on 3.24.0. Triggering schedule1. it is working fine.
* Apply the patch, started web server having end date feature. 
  * Triggered regular schedule2 without specifying endDate. Both sche1 and sche2 were running fine.
  * Triggered schedule3 by specifying endtime (5 minutes later). schedule3 was initially running. was able to observe schedule3 expire then.
  * Triggered schedule4 by specifying endtime (15 minutes later). Shutdown Webserver before schedule4 expires.

* restarted web server with end date implementation. Schedule3 never run. Both schedule1 and schedule2 and schdule4 were running well. Schedule4 expires at correct time.

* I also try rolling back web server to 3.24.0 (without this PR). Schedule 2 , 3 and 4 were not loaded due to lacking EndTimeChecker, but schedule1 was running good. Even though rolled back web server can not identify those latest schedules, the exceptions don't break and web server keeps working.

Though this PR looks big, but more than 80% code is related to testing and refactoring

### CODE REVIEW

**Comment by HappyRay** [comment]
"Schedule 2 , 3 and 4 were not loaded due to lacking EndTimeChecker" 
This makes rolling back the change impossible without user impact. Can you think of ways to mitigate? e.g. allow these flows to run as if the end date is not specified at least until we feel confident enough about the change.

**Comment by ameyamk** [comment]
End Date should be optional. We should have config/ policy from where we can enforce this. But as a default endDate should be optional. 
+1 to Ray.

File: azkaban-common/src/main/java/azkaban/scheduler/TriggerBasedScheduleLoader.java [hunk 1]
```diff
@@ -132,31 +128,24 @@ public void updateSchedule(Schedule s) throws ScheduleManagerException {
     }
   }
 
-  // TODO may need to add logic to filter out skip runs
+  // Todo kunkun-tang: Never used method, should be completely removed later.
   @Override
   public synchronized List<Schedule> loadSchedules()
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/scheduler/TriggerBasedScheduleLoader.java]: why not remove it now?

File: azkaban-common/src/main/java/azkaban/scheduler/TriggerBasedScheduleLoader.java [hunk 2]
```diff
@@ -167,15 +156,22 @@ private Schedule triggerToSchedule(Trigger t) throws ScheduleManagerException {
...
-              act.getExecutionOptions(), act.getSlaOptions(), ck.getCronExpression());
-      return s;
+    if (basicTimeChecker != null && act != null) {
+      return new Schedule(t.getTriggerId(),
+          act.getProjectId(),
+          act.getProjectName(),
+          act.getFlowName(),
+          t.getStatus().toString(),
+          basicTimeChecker.getFirstCheckTime(),
+          endTimeChecker == null?-1 : endTimeChecker.getEndCheckTime(),
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/scheduler/TriggerBasedScheduleLoader.java]: space

File: azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java [hunk 3]
```diff
@@ -300,7 +312,7 @@ private void onTriggerTrigger(Trigger t) throws TriggerManagerException {
       }
     }
 
-    private void onTriggerExpire(Trigger t) throws TriggerManagerException {
+    private void onTriggerPause(Trigger t) throws TriggerManagerException {
```
Comment (chengren311) [inline azkaban-common/src/main/java/azkaban/trigger/TriggerManager.java]: when will the trigger be expired if you change it to pause?

File: azkaban-web-server/src/main/java/azkaban/webapp/servlet/ScheduleServlet.java [hunk 4]
```diff
@@ -759,6 +767,14 @@ private void ajaxScheduleCronFlow(HttpServletRequest req,
       ret.put("error", e.getMessage());
     }
 
+    long endSchedTime = getLongParam(req, "endSchedTime", -1L);
+    try {
+      // Todo kunkun-tang: Need to verify if passed end time is valid.
```
Comment (chengren311) [inline azkaban-web-server/src/main/java/azkaban/webapp/servlet/ScheduleServlet.java]: ?

**Comment by kunkun-tang** [comment]
Closing this due to roll back issue. Reopen this PR at #1110

**Pull request closed without merging**